### PR TITLE
Allow multiple or missing Hive bucket files

### PR DESCRIPTION
**Pull request by v-jizhang**
Cherry-pick of trinodb/trino#822,
trinodb/trino#848 and
trinodb/trino#1375

Co-authored-by: David Phillips david@acz.org
Co-authored-by: Piotr Findeisen piotr.findeisen@gmail.com

```
== RELEASE NOTES ==

Hive Changes
* Allow multiple or missing Hive bucket files
This can be configured by using Hive Configuration Property ``hive.create-empty-bucket-files``.
Changes are also made to use Hive naming convention for bucket file names when computing bucket file name.
```

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/TestBackgroundHiveSplitLoader.java [hunk 1]
```diff
@@ -250,6 +252,23 @@ public void testCachingDirectoryLister()
                         "*,test_dbname.test_table"));
     }
 
+    @Test
+    public void testGetBucketNumber()
```
Comment (aweisberg) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestBackgroundHiveSplitLoader.java:256]: This belongs in `TestHiveWriterFactory`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 2]
```diff
@@ -829,7 +829,9 @@ private void testCreatePartitionedBucketedTableAsFewRows(
 
         assertUpdate(
                 // make sure that we will get one file per bucket regardless of writer count configured
-                getTableWriteTestingSession(optimizedPartitionUpdateSerializationEnabled),
+                Session.builder(getTableWriteTestingSession(optimizedPartitionUpdateSerializationEnabled))
+                        .setCatalogSessionProperty(catalog, "create_empty_bucket_files", String.valueOf(optimizedPartitionUpdateSerializationEnabled))
```
Comment (aweisberg) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java:833]: This is the wrong test parameter here. It's supposed to be `createEmpty` which wasn't added to the test method.

**Review by aweisberg — CHANGES_REQUESTED** [review]
Thank you for working on this!

Noticed some small things.

The commit messages don't follow our guidelines https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#example-commit-message Specifically the later commits don't have the original commit message and don't link to the PR in the body.

The release note isn't very detailed should link to the documentation for the new configuration options. I think it's also worth mentioning it changes the naming convention used by Presto for filenames just so people know it occurred.

I am not 100% sure we want to default to omitting empty bucket files. Generally I want to make sure this isn't going to break compatibility with systems that are expecting the existing output from Presto.

**Comment by v-jizhang** [comment]
> Thank you for working on this!
> 
> Noticed some small things.
> 
> The commit messages don't follow our guidelines https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#example-commit-message Specifically the later commits don't have the original commit message and don't link to the PR in the body.
> 
> The release note isn't very detailed should link to the documentation for the new configuration options. I think it's also worth mentioning it changes the naming convention used by Presto for filenames just so people know it occurred.
> 
> I am not 100% sure we want to default to omitting empty bucket files. Generally I want to make sure this isn't going to break compatibility with systems that are expecting the existing output from Presto.

I'll fix them and do another push once review is complete.

File: presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java [hunk 3]
```diff
@@ -3945,7 +3945,7 @@ protected void doCreateTable(SchemaTableName tableName, HiveStorageFormat storag
                     outputHandle.getLocationHandle().getTargetPath().toString(),
                     true);
             for (String filePath : listAllDataFiles(context, getStagingPathRoot(outputHandle))) {
-                assertTrue(new Path(filePath).getName().startsWith(getFilePrefix(outputHandle)));
+                assertTrue(new Path(filePath).getName().startsWith(session.getQueryId()));
```
[the remark this answers is not in the document]
Comment (v-jizhang) [reply presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:3948]: Just want to keep it as it is. :)

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 4]
```diff
@@ -89,7 +89,7 @@
     private HiveCompressionCodec orcCompressionCodec = HiveCompressionCodec.GZIP;
     private boolean respectTableFormat = true;
     private boolean immutablePartitions;
-    private boolean createEmptyBucketFiles ;
```
Comment (viczhang861) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: Nit:  try to fix trailing space in the original commit

**Review by viczhang861 — COMMENTED** [review]
1. Thank you very much for working on this.  
2. FYI, https://github.com/prestodb/presto/pull/15629/commits/e3f7edea3c18505753277fe163f8c827549d4bc3 improves the case for empty bucket file of temporary table,   with Hive version update to 3.0 already, we can combine the previous session property (for temporary table) and the one you added here, but let's do refactoring later when changes introduced in this PR are tested and stable.
3. Put Cherry-pick of Trino#848 as commit message and make commit title informative
4. You can cherry pick multiple PRs from Trino into one commit, whatever you think makes most sense.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestBackgroundHiveSplitLoader.java (hunk 1 above)
Comment (v-jizhang) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestBackgroundHiveSplitLoader.java]: Fixed. Thanks

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (hunk 2 above)
Comment (v-jizhang) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Added. Thanks

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java (hunk 4 above)
Comment (v-jizhang) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: Fixed, thanks.

File: presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (v-jizhang) [reply presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:3948]: @viczhang861 Commented below, I'll refactor it after this PR.

File: presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java [hunk 5]
```diff
@@ -354,53 +356,77 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
...
                     format("Hive table '%s' is corrupt. Found sub-directory in bucket directory for partition: %s",
-                            new SchemaTableName(table.getDatabaseName(), table.getTableName()),
+                            table.getSchemaTableName(),
                             partitionName));
         }
 
-        Map<Integer, HiveFileInfo> bucketToFileInfo = new HashMap<>();
+        ListMultimap<Integer, HiveFileInfo> bucketToFileInfo = ArrayListMultimap.create();
 
         if (!shouldCreateFilesForMissingBuckets(table, session)) {
```
[the remark this answers is not in the document]
Comment (v-jizhang) [reply presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java:365]: @viczhang861 Commented below, I'll refactor it after this PR.

**Review by aweisberg — APPROVED** [review]

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java (hunk 4 above)
Comment (highker) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: @v-jizhang, seems the change is still in the wrong commit?

**Review by highker — APPROVED** [review]

**Pull request merged**